contracts/other/notional/Join1155.sol

```diff
@@ -0,0 +1,122 @@
+    uint256 public immutable id;    // This ERC1155 Join only accepts one id from the ERC1155 token
+    uint256 public storedBalance;
+    uint256 public flashFeeFactor = FLASH_LOANS_DISABLED; // Fee on flash loans, as a percentage in fixed point with 18 decimals. Flash loans disabled by default.
+
+    constructor(address asset_, uint256 id_) {
+        asset = asset_;
+        id = id_;
+    }
+
+    /// @dev Advertising through ERC165 the available functions
+    function supportsInterface(bytes4 interfaceID) external view returns (bool) {
+        return  interfaceID == Join1155.supportsInterface.selector ||    // ERC-165 support (i.e. `bytes4(keccak256('supportsInterface(bytes4)'))`).
+                interfaceID == ERC1155TokenReceiver.onERC1155Received.selector ^ ERC1155TokenReceiver.onERC1155BatchReceived.selector;      // ERC-1155 `ERC1155TokenReceiver` support (i.e. `bytes4(keccak256("onERC1155Received(address,address,uint256,uint256,bytes)")) ^ bytes4(keccak256("onERC1155BatchReceived(address,address,uint256[],uint256[],bytes)"))`).
+    }
+
+    /// @dev Called by the sender after a transfer to verify it was received. Ensures only `id` tokens are received.
+    function onERC1155Received(
+        address _operator,
+        address _from,
+        uint256 _id,
+        uint256 _value,
+        bytes calldata _data
+    ) external override returns(bytes4) {
+        require (_id == id, "Token id not accepted");
+        return ERC1155TokenReceiver.onERC1155Received.selector;
+    }
+
+    /// @dev Called by the sender after a batch transfer to verify it was received. Ensures only `id` tokens are received.
+    function onERC1155BatchReceived(
+        address _operator,
+        address _from,
+        uint256[] calldata _ids,
+        uint256[] calldata _values,
+        bytes calldata _data
+    ) external override returns(bytes4) {
+        uint256 length = _ids.length;
+        for (uint256 i; i < length; ++i)
+            require (_ids[i] == id, "Token id not accepted");
+        return ERC1155TokenReceiver.onERC1155BatchReceived.selector;
+    }
+
+    /// @dev Take `amount` `asset` from `user` using `transferFrom`, minus any unaccounted `asset` in this contract.
+    function join(address user, uint128 amount)
+        external override
+        auth
+        returns (uint128)
+    {
+        return _join(user, amount);
+    }
+
+    /// @dev Take `amount` `asset` from `user` using `transferFrom`, minus any unaccounted `asset` in this contract.
+    function _join(address user, uint128 amount)
+        internal
+        returns (uint128)
+    {
+        ERC1155 token = ERC1155(asset);
+        uint256 _storedBalance = storedBalance;
+        uint256 available = token.balanceOf(address(this), id) - _storedBalance; // Fine to panic if this underflows
+        storedBalance = _storedBalance + amount;
+        unchecked { if (available < amount) token.safeTransferFrom(user, address(this), id, amount - available, ""); }
```

Comment:
we could prob include the line above in this checked block

Reply:
Done, including comment as to why this is safe